File: frontend/src/services/departmentService.js
import api from "./api";

export const getDepartments = async () => {
  const response = await api.get("/departments/");
  return response.data;
};

export const getDepartment = async (departmentId) => {
  const response = await api.get(`/departments/${departmentId}`);
  return response.data;
};

export const createDepartment = async (data) => {
  const response = await api.post("/departments/", data);
  return response.data;
};

export const updateDepartment = async (departmentId, data) => {
  const response = await api.put(
    `/departments/${departmentId}`,
    data
  );
  return response.data;
};

export const deleteDepartment = async (departmentId) => {
  const response = await api.delete(
    `/departments/${departmentId}`
  );
  return response.data;
};

File: backend/app/services/department_service.py
from sqlalchemy.orm import Session

from app.models.department import Department
from app.repositories.department_repository import (
    create_department,
    delete_department,
    get_department,
    get_departments,
    update_department,
)
from app.schemas.department import DepartmentCreate


def create_department_service(
    db: Session,
    data: DepartmentCreate,
) -> Department:
    department = Department(**data.model_dump())
    return create_department(db, department)


def get_departments_service(
    db: Session,
) -> list[Department]:
    return get_departments(db)


def get_department_service(
    db: Session,
    department_id: int,
) -> Department | None:
    return get_department(db, department_id)


def update_department_service(
    db: Session,
    department_id: int,
    data: DepartmentCreate,
) -> Department | None:
    department = get_department(db, department_id)

    if not department:
        return None

    for field, value in data.model_dump().items():
        setattr(department, field, value)

    return update_department(db, department)


def delete_department_service(
    db: Session,
    department_id: int,
) -> bool:
    department = get_department(db, department_id)

    if not department:
        return False

    delete_department(db, department)
    return True